[devgony] step3 score calculation

devgony opened the pull request:
늦어서 죄송합니다🙇‍♂️ 
사다리를 병행하던 도중 볼링 2단계에서 제가 구현했던 코드가 너무 지저분하다는 사실을 깨달았습니다. ㅠㅠ
볼링3단계는 처음부터 다시 해야겠다는 생각이 들어 여러 시도를 하다보니 많이 늦었습니다.
잘 부탁드립니다.

devgony committed:
step2: refactor: create factory method of Scores for Player

devgony committed:
step2: refactor: 컬렉션을 리턴해줄때는 불변으로 리턴 (PlayFrames)

devgony committed:
step2: chore: 주석 제거

devgony committed:
refactor: print 로직을 Output 으로 이동

devgony committed:
feat: 누계 점수 기능 구현
    -  Subtotal class 생성
      - 인자로 이전 subtotal을 받아 합산
      - 생성 시 strike, miss, spare 경우에 따라 waiting 설정
    -  Scores -> Frames 로 이름 변경하고 subtotals 추가
    -  plays 내에서 subtotal 생성하고 evaluateBonus 로직 추가
    -  10 프레임 이후에는 handleLast 에서 strike, spare 인 경우 한번 더 투구: playBonus
      - 이외의 경우에는 subtotal.setLast 로 기다리지 않아도 됌을 명시하고 최종 점수 출력

devgony committed:
test: domain test 추가
  - Player
  - Frames
  - Scores
  - Subtotal

devgony committed:
temp1: Refactoring 가능 class 확인

devgony committed:
temp2-refactoring

devgony committed:
feat: Refactoring from scratch - implement frame

devgony committed:
feat: Frames first class collection

devgony committed:
feat: create frames with linked list

devgony committed:
chore: rename tests

devgony committed:
feat: add Scores, Frame, Frames but pitches all at once

devgony committed:
feat: updateSubtotal accumulate previous subtotal

devgony committed:
feat: implement enum State, WAIT_TWICE

devgony committed:
fix: pitch can update both of previous and current

devgony committed:
feat: split FrameGeneral and FrameLast with interface

devgony committed:
feat: FrameLast with remaingTry3

devgony committed:
feat: FrameLast.pitch with scores.evaluateLastBonus can handle 10th frame

devgony committed:
style: wider record format

devgony committed:
feat: toSymbol can draw symbol for each numbers

devgony committed:
fix: last frame's state should be INIT or DONE

devgony committed:
chore: remove comments

devgony committed:
test: add test cases for Frames

devgony committed:
test: add test for Subtotal

devgony committed:
test: add test case for State

devgony committed:
feat: add class Player to scan name

devgony committed:
refactor: rename, rearrange package

devgony committed:
refactor: rename package from bowling to bowling_step3

devgony commented on src/main/java/bowling_step3/domain/FrameLast.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.List;
+import java.util.Objects;
+
+public class FrameLast implements Frame {
```
Frame 인터페이스를 implement 하여 FrameLast 를 구현 하였습니다.
여기서 중복되는 메서드는 쓰고싶지 않은데 혹시 방법이 있을까요?
아 extends FrameGeneral 을 하고 필요한 메서드만 override 시도해 보았는데 생성자가 FrameGeneral의 생성자를 바라보는것처럼 동작하여 제대로 구현하지못했습니다.

javajigi replied:
일반적으로 구현 클래스와 인터페이스 사이에 추상 클래스를 추가해 중복되는 코드를 제거한다.
다시 한번 도전해 볼 것을 추천한다.

devgony commented on src/main/java/bowling_step3/domain/Frames.java:
```diff
@@ -0,0 +1,71 @@
+        return this.frames.get(0);
+    }
+
+    public Frame next(Frame frame) {
```
next 를 Frame에서 가능하게 하려했으나 어차피 10개의 frame은 List로 구성되어있으므로 하나의 멤버변수를 절약하기 위해 Frames에서 next를 구현하였습니다.
반대로 저의 방식은 내부적으로는 index를 통해 get 하는 방식인데 이렇게 구현해도 괜찮을까요?

javajigi replied:
지금처럼 구현해도 되지 않을까?
지금과 같이 구현하다 뭔가 찜찜함이 있다면 그 시점에 Frame이 next에 접근하는 방식으로 구현해도 좋겠다.

devgony commented on src/main/java/bowling_step3/domain/FrameGeneral.java:
```diff
@@ -0,0 +1,90 @@
+        this.scores = scores;
+        if (scores.done()) {
+            State state = evaluateState(scores);
+            this.subtotal = new Subtotal(state, this.subtotal.value() + scores.sum());
```
L55: `this.scores = scores;` 나 
L65: `            this.subtotal = new Subtotal(state, this.subtotal.value() + scores.sum());` 처럼 실제로는 scores와 subtotal 을 set 하는 형태로 진행되고있습니다.
immutable을 위해 매번 Frame 을 새로 생성하는게 더 나을까요?

javajigi replied:
immutable을 지향하기 위해 최대한 노력해보고 방법을 찾지 못할 경우 mutable로 구현하는 식으로 접근할 것을 추천한다.

devgony reviewed (commented):
처음부터 다시 시작하는동안 모호했던 부분을 질문 남겨드립니다.
감사합니다. 🙇‍♂️

devgony committed:
feat: show frame index

devgony committed:
docs: add result

devgony committed:
refactor: rename Scores.remainingTry to remainingPitch

devgony committed:
docs: add second implement

devgony commented:
사전 요건정의가 부족했던 것 같아 README에 항목을 추가하였습니다. 감사합니다.

bingbingpa commented:
안녕하세요! 미션 기간이 끝났는데도 열심히 하는 모습 보기 좋네요! 
퇴근하고 리뷰를 하려고 했으나 오늘 퇴근이 늦었습니다 ㅜ
아마도 내일도 퇴근이 늦어질거 같은데 주말까지는 리뷰를 하도록 하겠습니다.

devgony commented:
아닙니다 ㅠ 봐주시는것만으로도 감사하죠ㅠ 저도 요즘 주말 출근이 계쏙되고있어서 천천히 봐주세요!

javajigi commented on src/main/java/bowling_step3/domain/Frame.java:
```diff
@@ -0,0 +1,17 @@
+    void updateNextSubtotal(Subtotal subtotal);
+
+    List<Integer> scores();
+}
```
인터페이스 추가 👍

javajigi commented on src/main/java/bowling_step3/domain/FrameGeneral.java:
```diff
@@ -0,0 +1,90 @@
+        pitch(scores, frames);
+    }
+
+    public void pitchRandom(Frames frames) {
```
manual과 random으로 분리한 이유가 있을까?

javajigi commented on src/main/java/bowling_step3/domain/FrameGeneral.java:
```diff
@@ -0,0 +1,90 @@
+            frames.prev(this).subtotal().accumulateBonus(this.scores.lastScore());
+            this.subtotal = new Subtotal(this.subtotal.state(), this.subtotal.value() + this.scores.lastScore());
+        }
+    }
```
투구를 할 때마다 점수를 계산하기보다 투구할 때는 투구 정보만 관리하고, getScore()와 같은 메서드를 호출하는 시점에 점수를 계산하도록 구현하는 것은 어떨까?

javajigi commented on src/main/java/bowling_step3/domain/Subtotal.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Override
+    public String
```
컨벤션 위반

javajigi commented on src/main/java/bowling_step3/domain/Subtotal.java:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Subtotal subtotal = (Subtotal) o;
+        return value == subtotal.value && state == subtotal.state;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(state, value);
+    }
```
일반적으로 equals, hashCode는 클래스의 제일 아래에 구현하는 것을 컨벤션으로 가진다.

javajigi reviewed (changes requested):
이 미션의 3단계 구현하기 정말 쉽지 않죠?
쉽지 않은 도전이었을텐데 수고하셨네요. 👍 

지금도 구현 잘 했는데요. 지금보다 더 작은 단위로 객체를 나눠보면 좋을 것 같아요. 설계하다 힘들면 강의 문서에 있는 힌트도 참고해 보셔도 좋을 것 같아요.

devgony committed:
refactor: equals, hashCode 는 클래스의 제일 아래에 구현

devgony committed:
refactor: 인터페이스와 구현 클래스 사이에 추상 클래스 추가하여 중복 제거_1: FrameLast

devgony committed:
refactor: 인터페이스와 구현 클래스 사이에 추상 클래스 추가하여 중복 제거_2: FrameGeneral

devgony committed:
refactor: pitch 에서 투구할 때는 투구 정보만 관리, score, subtotal 계산은 별도 분리
  - rename pitchManual to playManual
  - rename pitchRandom to pitchRandom
  - split to updateScore
  - split to updateSubtotal

devgony committed:
docs: add questions to unsolved list

devgony commented on src/main/java/bowling_step3/domain/FrameMutual.java:
```diff
@@ -0,0 +1,83 @@
+    public void playRandom(Frames frames) {
+        Scores scores = this.scores.pitchRandom();
+        updateScore(scores, frames);
+        updateSubtotal(frames);
```
처음에는 pitch, updateScore, updateSubtotal 을 모두 각각 ControllerApp 에서 수행하는 것을 생각했지만 그렇게 되면 controller에 내용이 많아지게 될것으로 보여 playRandom 이라는 메서드로 묶었습니다. 이렇게 해도 괜찮을까요?
혹은 이렇게 3개의 로직을 묶어주는 메서드는 필요하다면 컨트롤러로 분리하는게 좋을지 궁금합니다.

javajigi replied:
지난 번 피드백에도 있었듯이 매 투구를 할 때마다 점수를 변경하는 구조보다는 투구할 때는 투구 정보만을 저장하고 저장한 정보를 바탕으로 getScore()와 같은 메서드를 호출할 때 점수를 계산하는 방식으로 구현하는 것은 어떨까?

devgony commented on src/main/java/bowling_step3/domain/FrameMutual.java:
```diff
@@ -0,0 +1,83 @@
+        this(new Scores(), new Subtotal());
+    }
+
+    public void playManual(int numPins, Frames frames) {
```
프레임 단위의 투구에 대한 테스트를 위해 playManual 을 별도로 만들었었습니다.
score 도메인에서 이미 random 과 manual 에 대한 테스트 가 되어 있으므로 불필요한 메서드 인지 궁금합니다.

javajigi replied:
테스트를 위해 메서드를 추가하는 것은 추천하지 않는다.
지난 번 피드백에도 있지만 playManual과 playRandom이 같이 존재하는 것이 바람직해보이지 않는다.
play하나로 구현하고 테스트도 가능한 구조를 찾아보면 좋겠다.

devgony reviewed (commented):
이어서 리뷰를 해주셔서 감사드립니다 포비님!
피드백 주시는 대로 최대한 빨리 반영해서 마무리할 수 있도록 하겠습니다. 🙇‍♂️

javajigi commented on src/main/java/bowling_step3/domain/FrameMutual.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.List;
+import java.util.Objects;
+
+public abstract class FrameMutual implements Frame {
```
추상 클래스 추가를 통해 중복 제거 👍

javajigi commented on src/main/java/bowling_step3/domain/FrameMutual.java:
```diff
@@ -0,0 +1,83 @@
+
+public abstract class FrameMutual implements Frame {
+    Scores scores;
+    Subtotal subtotal;
```
인스턴스 변수는 특별한 이유가 없는 한 private으로 구현해 본다.
 private으로 구현할 경우 많은 제약이 따른다.
이런 제약이 따르는 상태에서 구현하는 연습을 해볼 것을 추천한다.

javajigi commented on src/test/java/bowling_step3/domain/FrameGeneralTest.java:
```diff
@@ -37,95 +37,95 @@ public class FrameGeneralTest {
-        fourth.pitchManual(8, frames);
+        fourth.playManual(8, frames);
         assertThat(fourth.subtotal()).isEqualTo(new Subtotal(State.INIT, 47));
     }
```
단위 테스트 단위가 너무 크다.
더 작은 단위로 단위 테스트를 분리해 보는 도전을 해보면 어떨까?

javajigi reviewed (changes requested):
추상 클래스를 추가해 리팩토링한 점 👍 

추가로 피드백 남겼으니 확인해 보시고요.
지금 단계에서 더 이상의 개선이 없다면 강의 문서의 "볼링 점수판 리팩토링 힌트"와 https://github.com/next-step/java-bowling/tree/javajigi 에 제가 구현한 코드도 있는데요. 
제 코드도 한번 보시고 리팩토링할 때 반영할 부분이 있는지 검토해 보는 것도 학습 측면에서 좋겠네요.

devgony committed:
- refactor: FrameMutual.scores, subtotal private 으로 변경
  - rename updateScore to evaluateScore

devgony committed:
test: add test for Scores to refactor

devgony committed:
test: add test for FrameGeneral to refactor

devgony committed:
test: add test for FrameLast_1

devgony committed:
test: add test for FrameLast_2

devgony committed:
feat: add Frame.done, getScore. todo: Frame.calculateAdditionalScore

devgony committed:
refactor: impl calculateAdditionalScore, add FrameLastTest

devgony committed:
test: add E2E test, todo: solve frame index

devgony committed:
temp

devgony committed:
fix: implement Frame with next

devgony committed:
test: FrameGeneralTest take1

devgony committed:
test: FrameLastTest take1

devgony committed:
test: FrameLastTest take2

devgony committed:
test: FrameLastTest done

devgony committed:
test: E2ETest take1

devgony committed:
feat: add Subtotals

devgony committed:
todo: calculateAdditionalScore for nextFrame

devgony committed:
feat: implement subtotal take1
todo: should handle accumulation

devgony committed:
feat: implement subtotal take2

devgony committed:
feat: subtotal take3- can accumulate

devgony committed:
feat: printSubtotals take1

devgony committed:
feat: printSubtotals take2
test: printSubtotalsWithFives

devgony committed:
fix: handle perfect game take1

devgony committed:
fix: handle perfect game take2, consecutive 10 gives 30 subtotal

devgony committed:
fix: last miss should fetch subtotal

devgony committed:
refactor: randomePin with int

devgony committed:
chore: remove comments

devgony committed:
fix: lastFrame should finish within 3 pitches

devgony committed:
chore: remove redundants

devgony committed:
chore: remove comments

devgony committed:
docs: write todo, done

devgony committed:
style: rename directories

devgony committed:
test: renew SubtotalsTest

devgony committed:
refactor: remove Subtotal

devgony commented on src/main/java/bowling_step3/domain/FrameMutual.java:
```diff
@@ -0,0 +1,102 @@
+
+public abstract class FrameMutual implements Frame {
+    Scores scores;
+    private Frame nextFrame;
```
기존에는 nextFrame을 index를 구한다음에 가져오는 방식이었는데 확인해보니 객체의 주소값을 비교하는것이 아니라 값으로 비교했을때 가장 처음 찾는 index를도출하는 버그를 확인해서 결국 Frame이 nextFrame 을가지도록 수정하였습니다

devgony commented on src/test/java/bowling_step3/E2ETest.java:
```diff
@@ -0,0 +1,77 @@
+        Player player = new Player("tst");
+        Output.printFrames(10, frames, player);
+        Output.printSubtotals(subtotals, player);
+    }
```
기본적으로 포비님의 테스트 시나리오를 참고하여 해당 시나리오들은 모두 구현할 수 있도록 테스트를 수정하고 제 코드에서 다른부분들의 테스트를 추가하였습니다

devgony reviewed (commented):
구조 전체를 변경하다 보니 이번에도 늦었습니다.. 
잘 부탁드리겠습니다 🙇🏼

javajigi commented on README.md:
```diff
@@ -163,8 +163,208 @@ Frame 10
+- [x] play 하나로 구현하고 테스트도 가능한 구조 찾기
+    - 밖에서 randomPins 를 구한 후 파라미터로 입력
+- [x] play 시에는 play만 하여 scores를 구하고 이후에 createSubtotals 를 통해 중간누적점수를 구함
+- [x] 테스트를 작은 단위로 분리
```
todo list 작성하고 완료 여부 체크하는 과정 👍 
추가로 commit log도 작은 단위로 잘 작성한 점 👍

javajigi commented on src/main/java/bowling_step3/controller/BowlingApp.java:
```diff
@@ -0,0 +1,24 @@
+
+        Frames frames = Frames.create();
+        Frame frame =frames.first();
+        while (!(frame.next() == null && frame.done())) {
```
```suggestion
        while (!frame.finished()) {
```

위와 같이 frame이 게임 종료 여부를 판단하도록 하는 것은 어떨까?

javajigi commented on src/main/java/bowling_step3/domain/Frame.java:
```diff
@@ -0,0 +1,21 @@
+package bowling_step3.domain;
+
+public interface Frame {
```
인터페이스에 메서드 수가 많다.
인터페이스를 분리해 메서드 수를 분리할 수 있는 방법은 없을까?

javajigi commented on src/main/java/bowling_step3/domain/FrameLast.java:
```diff
@@ -0,0 +1,54 @@
+package bowling_step3.domain;
+
+public class FrameLast extends FrameMutual implements Frame {
```
```suggestion
public class FrameLast extends FrameMutual {
```

FrameMutual이 Frame 인터페이스를 구현하고 있다.
FrameMutual만 상속해도 되지 않을까?

javajigi replied:
추상 클래스임을 알 수 있도록 AbstractFrame과 같은 이름도 자주 사용함

javajigi commented on src/main/java/bowling_step3/domain/FrameMutual.java:
```diff
@@ -0,0 +1,102 @@
+    Scores scores;
+    private Frame nextFrame;
+
+    public FrameMutual(Scores scores, Frame nextFrame) {
```
이 생성자가 주 생성자인데 주 생성자는 일반적으로 생성자의 마지막에 구현하는 것이 관례임

javajigi commented on src/main/java/bowling_step3/domain/FrameMutual.java:
```diff
@@ -0,0 +1,102 @@
+package bowling_step3.domain;
+
+public abstract class FrameMutual implements Frame {
+    Scores scores;
```
인스턴스 변수를 private 접근 제어자로 변경하고 구현해 본다.

javajigi reviewed (approved):
전체적인 객체 설계(상속 포함), 클린 코드 구현, 단위 테스트 구현 💯 
특히 todo list 작성과 commit log가 인상적이네요.
단위 테스트도 작은 단위로 잘 구현했고요.

피드백이 좀 있기는 한데 다음 단계 도전할 때 같이 반영해 보면 좋겠어요.
힌트와 제 코드를 보면 State 패턴을 적용해 구현했는데요.
이번 기회에 State 패턴 적용해 구현해 보는 것도 좋은 연습이 될 겁니다.
